### PR TITLE
MAINT: Switch to GHA cibuildwheel

### PR DESCRIPTION
**Pull request by larsoner**
This is going to require some effort to get working probably, but figured I'd get the ball rolling.

1. Switch to more standard GHA cibuildwheel setup away from azure-pipelines
2. Move as much as possible to `pyproject.toml` instead of using env vars
3. Use native compiling for macos via `-13` (Intel) and `-14` (arm) runners, removing cross-compiling shims along the way
4. In PRs, build against latest NumPy release (no longer build against NumPy 2.1.0.dev).
5. On scheduled runs, build all wheels (hooray!) against scientific-python-nightly NumPy 2.1.0.dev, then upload the resulting wheels to scientific-python-nightly-wheels.
6. On PRs, build against NumPy 2.1.0.dev if `[pip-pre]` is in the commit message
7. On PRs, build all aarch64 variants if `[aarch64]` is in the commit message
8. On pushes, build all wheels against latest NumPy release. If a tag (i.e., a release), update the release with the wheels. If a `master` push, create a draft release with the artifacts. We probably just want this enabled *temporarily* until the action actually seems to work (might take a few PRs to tweak stuff) otherwise it might create a lot of draft releases, depending on your `h5py.__version__`.

Not totally sure I should have removed all the macOS stuff, hoping things will "just work" with less effort now that it's not based on cross-compilation :crossed_fingers: 

Todo:

- [x] Get macOS x86_64 working
- [x] Get macOS arm64 working
- [x] Get Linux x86_64 working
- [x] Get Windows AMD64 working
- [x] Check that caching of HDF5 works
- [x] Add Linux aarch64
- [x] Cross-check wheel artifacts against PyPI release (filenames and `h5py.get_config()`)
- [x] Add infrastructure for 3.13 nightlies but don't activate it

### CODE REVIEW

**Comment by neutrinoceros** [comment]
@larsoner Just so attention gets more easily focused on #2434, I'll edit the title and mark this as a draft.

**Comment by neutrinoceros** [comment]
@larsoner I see you undrafted  the PR but it doesn't seem to be working yet. Can you please clarify if you need a hand to get it working or if we should wait for you to stabilize it before reviewing ?

File: setup_build.py
Comment (aragilar) [inline setup_build.py:134]: This seems too general, if numpy has float128 support on a system, then this should work (It would be good to work out why MacOS doesn't give the correct result already).

Comment (aragilar) [reply setup_build.py:134]: It looks like this comes from https://github.com/h5py/h5py/pull/2065, which shouldn't be needed if we're not cross-compiling.

File: pyproject.toml
Comment (larsoner) [reply pyproject.toml:110]: I was trying to avoid changing behavior / use of env vars where possible, and the only place this seems to be set in `master` is in the macos wheels

https://github.com/h5py/h5py/blob/cc468545218b2465342f1f94f14103ac2848a682/azure-pipelines.yml#L52-L59

**Comment by larsoner** [comment]
> Can you please clarify if you need a hand to get it working or if we should wait for you to stabilize it before reviewing ?

Yeah sorry it was working back on https://github.com/h5py/h5py/pull/2444/commits/c575df4db3eefe7dbca3d05b4cc3aaed1f550505 then I couldn't resist the temptation to roll in aarch64 / https://github.com/h5py/h5py/pull/2434, so I'm doing that now. I'll set back to draft until it works!

File: pyproject.toml
Comment (aragilar) [reply pyproject.toml:111]: Interesting, we should probably check the reported configuration of all the wheels once this PR is ready, and ensure it's what should go onto PyPI.

**Comment by larsoner** [comment]
Argh it was working:

- [6eba64b](https://github.com/h5py/h5py/pull/2444/commits/6eba64b5d6b62f8ad7eec13dcd1f66bc984289df) uses `[pip-pre]` in the commit message and seemed to do the right thing (build against NumPy pip-pre)
- [a6593f1](https://github.com/h5py/h5py/pull/2444/commits/a6593f164851a47c4c6da1a4fbee3b911ede8cfa) uses `[aarch64]` so all aarch64 wheels get built. For speed I only run the `py12-test-deps` for aarch64 rather than the full `py$PYVER-test-deps,py$PYVER-test-mindeps,py$PYVER-test-deps-pre` (across all Python versions) on both PRs and pushes/tags, hope that's okay. It takes ~1h45m already, don't want to make it longer (can hopefully trust that it generalizes to other Pythons and deps etc.).

The second one failed because of https://github.com/tox-dev/filelock/issues/337 which hopefully will be reverted soon. Can't continue get green CIs until it does, but in the meantime I think this is ready for review!

File: h5py/tests/test_file.py
Comment (neutrinoceros) [inline h5py/tests/test_file.py:105]: I think this should be replaced by a `@pytest.mark.skipif` decorator. Or possibly use `pytest.skip` directly here so we still have the opportunity to fail the test in its early parts.

**Review by neutrinoceros — COMMENTED** [review]
thanks a lot for doing this. I don't have the bandwidth to review changes to `ci/*` in depth but here are a couple minor observations. Overall this is great !

File: .github/workflows/build_wheels.yml
Comment (larsoner) [reply .github/workflows/build_wheels.yml:149]: [Here](https://github.com/openmeeg/openmeeg/releases) is an example of what it looks like. I don't think it creates much noise because you can just have a `dev` release that gets updated.

At a minimum I would suggest keeping it *temporarily* at least, just to make sure it can actually create a release once this PR lands in `main`. If it fails then you get that failure now and can debug immediately rather than when you actually try to cut a release and it fails. That's an annoying situation!

File: .gitignore
Comment (larsoner) [reply .gitignore:37]: Based on how `.gitignore` works `/` is the *repo* root not filesystem root. Things like `.mypy_cache/` will match effectively `ROOT/.mypy_cache/`, `ROOT/whatever/.mypy_cache`, etc. Whereas `/cache/` will *only* match `ROOT/cache/`, not `ROOT/whatever/cache/`.

File: h5py/tests/test_file.py
Comment (larsoner) [reply h5py/tests/test_file.py:105]: > so we still have the opportunity to fail the test in its early parts.

FWIW I think `return` here also allows the test to fail in the preceding parts. For example if there's an error on line 98 or whatever, it will still be raised *before* this line is ever hit.

It's a weird situation because only *part* of the test gets skipped. The cleanest thing to do would probably be to split this into a separate test altogether actually, then the `skip` will be more meaningful / correct because when you see it, it would mean that all parts of the test were skipped. (If we just added it in place now, it would be misleading because some of the test will have run already.) So I'll do that! 👍

File: pyproject.toml
Comment (larsoner) [reply pyproject.toml:111]: I'm happy to cross-check first the set of artifacts produced in this PR with those on PyPI -- I can at least do `import`-based checks on macOS arm64, Linux x86_64, and Windows. What functions/commands/files should I check to see the configuration used to build the wheel?

Comment (larsoner) [reply pyproject.toml:111]: ... ahh, `h5py.get_config()` has some stuff I can use!

**Comment by larsoner** [comment]
> Interesting, we should probably check the reported configuration of all the wheels once this PR is ready, and ensure it's what should go onto PyPI.

As expected (thankfully!) the only difference in filenames between the set of wheels in [the latest PyPI release](https://pypi.org/project/h5py/#files) and [this PR](https://github.com/h5py/h5py/actions/runs/9491056331#artifacts) are that the PR has aarch64 wheels:

```
$ diff <(ls -C1 pr_wheels/ |sort) <(ls -C1 pypi_wheels/ |sort)
3d2
< h5py-3.11.0-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
8d6
< h5py-3.11.0-cp311-cp311-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
13d10
< h5py-3.11.0-cp312-cp312-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
18d14
< h5py-3.11.0-cp38-cp38-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
23d18
< h5py-3.11.0-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
```

And then looking at config options with:
```
python -c 'import h5py, pprint; pprint.pprint({key: getattr(h5py.get_config(), key) for key in sorted(dir(h5py.get_config())) if not key.startswith("_")})'
```
Things appear to be the same at least on:

<details>
<summary>macOS arm64 3.11</summary>

```
PyPI:
{'API_16': None,
 'API_18': None,
 'bool_names': (b'FALSE', b'TRUE'),
 'complex_names': ('r', 'i'),
 'default_file_mode': 'r',
 'direct_vfd': False,
 'mpi': False,
 'read_byte_strings': <h5py.h5.ByteStringContext object at 0x10147f3e0>,
 'ros3': False,
 'swmr_min_hdf5_version': (1, 9, 178),
 'track_order': False,
 'vds_min_hdf5_version': (1, 9, 233)}

e4e92a8:
{'API_16': None,
 'API_18': None,
 'bool_names': (b'FALSE', b'TRUE'),
 'complex_names': ('r', 'i'),
 'default_file_mode': 'r',
 'direct_vfd': False,
 'mpi': False,
 'read_byte_strings': <h5py.h5.ByteStringContext object at 0x1010c7fe0>,
 'ros3': False,
 'swmr_min_hdf5_version': (1, 9, 178),
 'track_order': False,
 'vds_min_hdf5_version': (1, 9, 233)}
```

</details>

<details>
<summary>Linux x86_64 3.12</summary>

```
PyPI
{'API_16': None,
 'API_18': None,
 'bool_names': (b'FALSE', b'TRUE'),
 'complex_names': ('r', 'i'),
 'default_file_mode': 'r',
 'direct_vfd': False,
 'mpi': False,
 'read_byte_strings': <h5py.h5.ByteStringContext object at 0x7a6df8ea8380>,
 'ros3': False,
 'swmr_min_hdf5_version': (1, 9, 178),
 'track_order': False,
 'vds_min_hdf5_version': (1, 9, 233)}

e4e92a8
{'API_16': None,
 'API_18': None,
 'bool_names': (b'FALSE', b'TRUE'),
 'complex_names': ('r', 'i'),
 'default_file_mode': 'r',
 'direct_vfd': False,
 'mpi': False,
 'read_byte_strings': <h5py.h5.ByteStringContext object at 0x75ca32380200>,
 'ros3': False,
 'swmr_min_hdf5_version': (1, 9, 178),
 'track_order': False,
 'vds_min_hdf5_version': (1, 9, 233)}
```

</details>

<details>
<summary>Windows 10 AMD64 3.11</summary>

```
PyPI
{'API_16': None,
 'API_18': None,
 'bool_names': (b'FALSE', b'TRUE'),
 'complex_names': ('r', 'i'),
 'default_file_mode': 'r',
 'direct_vfd': False,
 'mpi': False,
 'read_byte_strings': <h5py.h5.ByteStringContext object at 0x000001AF6C093390>,
 'ros3': False,
 'swmr_min_hdf5_version': (1, 9, 178),
 'track_order': False,
 'vds_min_hdf5_version': (1, 9, 233)}

e4e92a8
{'API_16': None,
 'API_18': None,
 'bool_names': (b'FALSE', b'TRUE'),
 'complex_names': ('r', 'i'),
 'default_file_mode': 'r',
 'direct_vfd': False,
 'mpi': False,
 'read_byte_strings': <h5py.h5.ByteStringContext object at 0x000002A116465B90>,
 'ros3': False,
 'swmr_min_hdf5_version': (1, 9, 178),
 'track_order': False,
 'vds_min_hdf5_version': (1, 9, 233)}
```

</details>

File: .github/workflows/build_wheels.yml
Comment (larsoner) [reply .github/workflows/build_wheels.yml:149]: To save some clicks, the `main` `dev` "release" shows up as a draft in the list of releases:

Screenshot from 2024-06-13 11-43-24

And it *doesn't* show up as the latest release on the GitHub landing page, which is nice:

Screenshot from 2024-06-13 11-43-35

Again, might not make sense as a permanent thing but at least temporarily it's fairly unobtrusive and would allow you to make sure wheels can get added as assets to a GitHub release, increasing the likelihood that it'll do the right thing when you do eventually cut a release.

File: h5py/tests/test_file.py
Comment (neutrinoceros) [reply h5py/tests/test_file.py]: > FWIW I think return here also allows the test to fail in the preceding parts.

that's right, I'm just worried that we might change the meaning of "✅" from "that test passed" to "at least it returned before it could fail !" 😸

File: h5py/tests/test_file.py
Comment (larsoner) [reply h5py/tests/test_file.py]: This is weird. With a debug `print` statement I can see the perms change from `33188` to `33024` and at least Python thinks it has become read-only:
```
>>> 33188 & 128
128
>>> 33024 & 128
0
```
I'll open an upstream issue and add a comment. But it's also possible this is a bug with h5py somehow or the custom build images used.

Comment (larsoner) [reply h5py/tests/test_file.py:106]: Okay removed the debug statements and linked to https://github.com/pypa/cibuildwheel/issues/1882

**Comment by larsoner** [comment]
Okay I tried enabling 3.13 but ran into https://github.com/h5py/hdf5-manylinux/issues/11 so left the code in place but didn't enable it for now. Docker is having a bad time so I'll force-push later to get things green again once it's working 🙄

**Comment by hoechenberger** [comment]
We're green ✅

**Comment by larsoner** [comment]
I took a stab at some commits to enable 3.13. It's doable but will take a bit of iteration, so I'll tackle it after this is merged.

**Comment by larsoner** [comment]
Okay updated to build 3.13 in Linux nightly builds to fix the merge conflict and catch up with #2447. If #2449 is merged then I'll work to catch up with that next.

image


No idea why macOS cancelled itself (?) but it was passing a few commits ago with no change to `azure-pipelines` so probably not relevant here.

**Comment by larsoner** [comment]
Friendly ping to consider this whenever someone gets a chance!

Also I realized that the `macos11` builds were cancelled because that image is gone so this now also bumps `azure` to a `macos12` image.

File: .travis.yml
Comment (takluyver) [inline .travis.yml:49]: There's a couple more aarch64-specific bits further down in this file, which could be cleaned up now that we're only using Travis for ppc64le.

**Comment by takluyver** [comment]
Thanks, this looks like a really nice simplification. @neutrinoceros has offered to review it in more detail, but the overall move seems like a good one to me.

I hope Github actions adds some Linux aarch64 machines, because going from 24 minutes to build for 5 Python versions on Travis to 29 minutes for one Python version feels like a step backwards. But Travis has been flaky enough recently that that seems like a price worth paying. And I suspect Github will gain aarch64 machines at some point - especially if lots of projects otherwise use an emulator.

I see from the diff that this still leaves some bits to run on Azure pipelines. It needn't be in this PR, but does it seem practical to move everything over to GHA so we have one fewer CI platform to manage? Do parallelism limits mean spreading it across two platforms is significantly quicker? Even if we don't go down this route, CI that just builds & tests is already simpler than dealing with job dependencies, artifacts etc., so this still feels like a step forwards. :slightly_smiling_face:

**Review by neutrinoceros — COMMENTED** [review]
In my opinion this is almost good to go, I just have a couple minor questions. I think this is the best I can do with limited knowledge of the existing setup and with the time I have on my hands, but I'm fairly confident that just seeing CI pass means that most defects are already resolved.

File: ci/cibw_before_all_macos.sh
Comment (larsoner) [reply ci/cibw_before_all_macos.sh]: Did not seem to work:
```
  delocate.libsana.DelocationError: Library dependencies do not satisfy target MacOS version 11.0:
  /private/var/folders/xz/lz5thm6s3vb1vdkk77vmkgbr0000gn/T/tmpxrqwmqvd/wheel/h5py/.dylibs/libhdf5.200.dylib has a minimum target of 14.0
```

Comment (neutrinoceros) [reply ci/cibw_before_all_macos.sh]: Not a big deal. Thanks for trying anyway !

Comment (larsoner) [reply ci/cibw_before_all_macos.sh]: ... and now I remember why. In principle this should be a `before_all` step, but in order to cache it properly I thought I needed to run it separately. But I don't think I need to actually... I'll try rolling it into regular cibuildwheel `before_all` and see if it works.

Comment (larsoner) [reply ci/cibw_before_all_macos.sh]: Okay I think the sticking point here is that we need to set a bunch of env vars when getting HDF5. On macOS it's LD_LIBRARY_PATH, PKG_CONFIG_PATH, etc. On Windows it's CL, LINK, ZLIB_ROOT, etc. If we do these steps *before* cibuildwheel is invoked it's easy -- we just `echo "ZLIB_ROOT=$ZLIB_ROOT" | tee -a $GITHUB_ENV` for example.

If we're *already in the cibuildwheel step* this does not work -- it doesn't add the env vars to the currently running step's (cibuildwheel step's) environment. And I don't think you can `source` the scripts in the `before_all` to get env vars back to the cibuildwheel step AFAIK.

One way around this would be to define `CIBUILDWHEEL_ENVIRONMENT_MACOS` for example in `pyproject.toml` or in `build_wheels.yml` to set the vars, but I think this is overall a step back -- it pushes setting their values to a YAML or TOML file whereas currently they're set in a script where they're defined.

So I'm going to revert things a bit and fix the macOS build, and mention why it's done this way.

Comment (neutrinoceros) [reply ci/cibw_before_all_macos.sh:20]: excellent, thank you !

**Review by neutrinoceros — APPROVED** [review]

**Comment by larsoner** [comment]
Friendly ping, should we merge this and continue in follow-up PRs?

**Comment by neutrinoceros** [comment]
I think it's reasonable to merge soon, yes. If we don't hear from @takluyver or @tacaswell by the end of the week I'll push the button.

**Comment by aragilar** [comment]
This looks good to me as well. The only minor concern I have is around arm64 wheel (given previous "fun" with float types being misdetected), but given issues with travis, as long as users aren't seeing any issues with it, I think we can can use emulation until GitHub adds native runners (i.e. don't block on this).

Did we create an issue for cibuildwheel having the broken test?

**Comment by larsoner** [comment]
> Did we create an issue for cibuildwheel having the broken test?

I opened https://github.com/pypa/cibuildwheel/issues/1882, could open an issue here if it helps.

**Comment by neutrinoceros** [comment]
Ok let's push that button 🙈

**Comment by takluyver** [comment]
Thanks @larsoner for your work on this - I've just released h5py 3.12, and it was definitely easier having all the wheels in one place. I used the `gh` command line to download them all, and then upload to PyPI.

Now that the wheels are all built on GHA, I guess it would be relatively simple to also build an sdist, and then upload everything to PyPI on a release using trusted publishing.

**Comment by larsoner** [comment]
Awesome! Yes trusted publishing is how I've been doing things for some months now and the process and documentation for doing it was pretty straightforward 👍

**Pull request merged**
